File: baZiApp/src/types/bazi.ts
// 四柱八字核心类型定义

// 天干
export type TianGan = '甲' | '乙' | '丙' | '丁' | '戊' | '己' | '庚' | '辛' | '壬' | '癸';

// 地支
export type DiZhi = '子' | '丑' | '寅' | '卯' | '辰' | '巳' | '午' | '未' | '申' | '酉' | '戌' | '亥';

// 五行
export type WuXing = '金' | '木' | '水' | '火' | '土';

// 十神
export type ShiShen = '比肩' | '劫财' | '食神' | '伤官' | '正财' | '偏财' | '正官' | '七杀' | '正印' | '偏印' | '日主';

// 阴阳
export type YinYang = '阳' | '阴';

// 单柱（天干+地支）
export interface Zhu {
  tiangan: TianGan;
  dizhi: DiZhi;
  shishen?: ShiShen; // 十神关系
  canggan?: TianGan[]; // 藏干
}

// 四柱八字
export interface SiZhu {
  year: Zhu;  // 年柱
  month: Zhu; // 月柱
  day: Zhu;   // 日柱
  hour: Zhu;  // 时柱
}

// 五行统计
export interface WuXingCount {
  element: WuXing;
  count: number;
  percentage: number;
}

// 排盘输入参数
export interface PaiPanInput {
  name: string;
  gender: '男' | '女';
  birthDate: string; // YYYY-MM-DD
  birthTime: string; // HH:mm
  birthPlace: string; // 出生地
  longitude?: number; // 经度
  latitude?: number;  // 纬度
  paipanMode?: 'simple' | 'full'; // 排盘模式：simple=普通式, full=全排
}

// 排盘结果
export interface PaiPanResult {
  input: PaiPanInput;
  sizhu: SiZhu;
  wuxing: WuXingCount[];
  dayMaster: TianGan; // 日主
  dayMasterStrength: '强' | '弱' | '中和';
  analysis: AnalysisReport;
  // 全排专属字段（普通式为空）
  taiyuan?: { tiangan: TianGan; dizhi: DiZhi };
  minggong?: { tiangan: TianGan; dizhi: DiZhi };
  shengong?: { tiangan: TianGan; dizhi: DiZhi };
  dayun?: DaYun[];
  liunian?: LiuNian[];
  xiaoyun?: XiaoYun;
  tongyun?: TongYun[];
  shensha?: ShenShaResult;
  emptyPosition?: string[];
  // 起运信息（扩展字段）
  qiyun?: QiYunInfo;
}

// 起运信息
export interface QiYunInfo {
  forward: boolean;
  direction: string;  // 顺行/逆行
  startAge: number;    // 实岁
  startAgeXu: number;  // 虚岁
  qiyunDays: number;   // 起运天数
  qiyunMonths: number; // 起运月数
  refTerm: string;     // 参考节气
  jiaoyunYear: number; // 交运年份
  jiaoyunMonth: number; // 交运月份
}

// 分析解读报告
export interface AnalysisReport {
  overview: string; // 总体概述
  dayMasterAnalysis: string; // 日主分析
  wuxingAnalysis: string; // 五行分析
  shishenAnalysis: string; // 十神分析
  wealth: string; // 财运
  career: string; // 事业
  health: string; // 健康
  marriage: string; // 婚姻
  education: string; // 学业
  suggestions: string[]; // 建议
}

// 大运
export interface DaYun {
  index: number; // 第几步大运（从0开始）
  startAge: number; // 起运年龄
  endAge: number; // 截止年龄
  tiangan: TianGan; // 天干
  dizhi: DiZhi; // 地支
  shishen: ShiShen; // 十神
  startYear: number; // 起始年份
  endYear: number; // 结束年份
}

// 流年
export interface LiuNian {
  year: number; // 年份
  tiangan: TianGan; // 天干
  dizhi: DiZhi; // 地支
  shishen: ShiShen; // 十神（相对于日主）
  dayMasterRelation: string; // 与日主的关系
}

// 小运
export interface XiaoYun {
  year: number;
  tiangan: TianGan;
  dizhi: DiZhi;
}

// 童运
export interface TongYun {
  age: number; // 年龄（1-12岁）
  tiangan: TianGan;
  dizhi: DiZhi;
  shishen: ShiShen;
}

// 神煞
export interface ShenSha {
  name: string; // 神煞名称
  position: '年柱' | '月柱' | '日柱' | '时柱' | '年干' | '月干' | '日干' | '时干' | '年支' | '月支' | '日支' | '时支';
  description?: string;
}

// 神煞结果
export interface ShenShaResult {
  yearPillar: ShenSha[]; // 年柱神煞
  monthPillar: ShenSha[]; // 月柱神煞
  dayPillar: ShenSha[]; // 日柱神煞
  hourPillar: ShenSha[]; // 时柱神煞
  all: ShenSha[]; // 所有神煞
}

// 旺衰
export type WangShuai = '长生' | '沐浴' | '冠带' | '临官' | '帝旺' | '衰' | '病' | '死' | '墓' | '绝' | '胎' | '养';

// 纳音
export interface NaYin {
  element: string; // 纳音五行
  description: string; // 纳音描述
}

// 扩展的四柱（包含更多信息）
export interface ZhuDetail extends Zhu {
  wangShuai?: WangShuai; // 旺衰
  nayin?: string; // 纳音
  shishenDetail?: {
    tiangan: ShiShen;
    dizhi: ShiShen;
  };
}

// 扩展的排盘结果
export interface PaiPanResultExtended extends PaiPanResult {
  sizhuDetail: {
    year: ZhuDetail;
    month: ZhuDetail;
    day: ZhuDetail;
    hour: ZhuDetail;
  };
  dayun: DaYun[]; // 大运
  shenSha: ShenShaResult; // 神煞
  xiaoYun?: XiaoYun; // 当前小运
  tongYun?: TongYun[]; // 童运
  emptyPosition: string[]; // 空亡位置
}

// API响应
export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: string;
  message?: string;
}

File: backend/app/schemas/bazi.py
# app/schemas/bazi.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import date, time


class BirthInfo(BaseModel):
    """出生信息输入"""
    name: Optional[str] = Field(None, description="姓名（可选）")
    gender: Optional[str] = Field(None, description="性别：male/female")
    birth_date: str = Field(..., description="出生日期 YYYY-MM-DD")
    birth_time: str = Field(..., description="出生时间 HH:MM（北京时间）")
    city: Optional[str] = Field(None, description="出生城市名称")
    latitude: Optional[float] = Field(None, ge=-90, le=90, description="纬度")
    longitude: Optional[float] = Field(None, ge=-180, le=180, description="经度")
    use_true_solar: bool = Field(False, description="是否使用真太阳时")


class BaziPillar(BaseModel):
    """单柱信息"""
    stem: str = Field(..., description="天干")
    branch: str = Field(..., description="地支")
    hidden_stems: List[str] = Field(default_factory=list, description="地支藏干")
    wuxing: str = Field(..., description="五行")


class ShishenItem(BaseModel):
    """十神项"""
    stem: str = Field(..., description="天干")
    target: str = Field(..., description="作用天干")
    shishen: str = Field(..., description="十神名称")
    wuxing: str = Field(..., description="五行属性")


class AnalysisReportSchema(BaseModel):
    """命理解读报告"""
    overview: str = Field(..., description="总体概述")
    dayMasterAnalysis: str = Field(..., description="日主分析")
    wuxingAnalysis: str = Field(..., description="五行分析")
    shishenAnalysis: str = Field(..., description="十神分析")
    wealth: str = Field(..., description="财运分析")
    career: str = Field(..., description="事业分析")
    health: str = Field(..., description="健康分析")
    marriage: str = Field(..., description="婚姻分析")
    education: str = Field(..., description="学业分析")
    suggestions: List[str] = Field(default_factory=list, description="改善建议")


class BaziResult(BaseModel):
    """八字排盘结果"""
    year_pillar: BaziPillar = Field(..., description="年柱")
    month_pillar: BaziPillar = Field(..., description="月柱")
    day_pillar: BaziPillar = Field(..., description="日柱")
    hour_pillar: BaziPillar = Field(..., description="时柱")
    day_master: str = Field(..., description="日主（日干）")
    day_master_wuxing: str = Field(..., description="日主五行")
    day_master_yinyang: str = Field(..., description="日主阴阳")
    year_shishen: Optional[ShishenItem] = Field(None, description="年柱十神")
    month_shishen: ShishenItem = Field(..., description="月柱十神")
    hour_shishen: Optional[ShishenItem] = Field(None, description="时柱十神")
    # 纳音
    year_nayin: str = Field(..., description="年柱纳音")
    month_nayin: str = Field(..., description="月柱纳音")
    day_nayin: str = Field(..., description="日柱纳音")
    hour_nayin: str = Field(..., description="时柱纳音")
    # 胎元、命宫、身宫（可选）
    taiyuan: Optional[BaziPillar] = Field(None, description="胎元")
    mggong: Optional[BaziPillar] = Field(None, description="命宫")
    shengong: Optional[BaziPillar] = Field(None, description="身宫")
    # 真太阳时信息
    true_solar_time: Optional[str] = Field(None, description="真太阳时间 HH:MM")
    true_solar_adjustment: Optional[int] = Field(None, description="调整分钟数")
    # 大运
    dayun: Optional[dict] = Field(None, description="大运信息")
    # 神煞
    shensha: Optional[dict] = Field(None, description="神煞信息")
    # 旺衰
    wangshuai: Optional[dict] = Field(None, description="旺衰信息")
    # 格局判定（子平真诠核心）
    geju: Optional[dict] = Field(None, description="格局判定结果（含geju/chengge/desc等）")
    # 用神/忌神
    yongshen: Optional[dict] = Field(None, description="用神/忌神信息（含yongshen/jishen/tiaohou/tongguan/type/strength_auto/desc）")
    # 调候用神（穷通宝鉴）
    tiaohou: Optional[dict] = Field(None, description="调候用神结果（含tiaohou_yongshen/hannuan/zaoshi/desc）")
    # 命理解读报告
    analysis_report: Optional[AnalysisReportSchema] = Field(None, description="命理解读报告")


class BaziRequest(BaseModel):
    """八字排盘请求"""
    birth_info: BirthInfo


class BaziResponse(BaseModel):
    """八字排盘响应"""
    success: bool
    data: Optional[BaziResult] = None
    message: Optional[str] = None
